src/main/scala/pureconfig/ConfigFieldMapping.scala

```diff
@@ -0,0 +1,24 @@
+package pureconfig
+
+trait ConfigFieldMapping[T] {
```

Comment:
The current implementation in this PR assumes that you have an implicit `ConfigFieldMapping` in scope that tells you how fields of classes should map to fields in the config. There's always one available that maps the fields using the identity function. This allows you to have different mappings for different types. For example:

```scala
import pureconfig._
import pureconfig.syntax._
import com.typesafe.config._

case class ConfWithCamelCaseInner(thisIsAnInt: Int, thisIsAnotherInt: Int)
case class ConfWithCamelCase(camelCaseInt: Int, camelCaseString: String, camelCaseConf: ConfWithCamelCaseInner)

implicit val conv = new WordDelimiterConfigFieldMapping[ConfWithCamelCase](CamelCaseWordDelimiter, HyphenWordDelimiter)

val conf = ConfigFactory.parseString("""{
  camel-case-int = 1
  camel-case-string = "bar"
  camel-case-conf {
    thisIsAnInt = 3
    thisIsAnotherInt = 10
  }
}""")

conf.to[ConfWithCamelCase]
// returns Success(ConfWithCamelCase(1,bar,ConfWithCamelCaseInner(3,10)))
```

I like to have this flexibility and I have a use case for it in a project. However, I understand that using implicit options is error-prone and a bit magical for the end user. What I was suggesting in my last comment was to have a method to create a `ConfigConvert` from a `ConfigFieldMapping`, so that we no longer have to provide it implicitly (having an implicit `ConfigConvert` in scope is already a reasonable compromise of this library). This would still allow having different mappings for different types, but would simplify the interface:

```scala
import pureconfig._
import pureconfig.syntax._
import com.typesafe.config._

case class ConfWithCamelCaseInner(thisIsAnInt: Int, thisIsAnotherInt: Int)
case class ConfWithCamelCase(camelCaseInt: Int, camelCaseString: String, camelCaseConf: ConfWithCamelCaseInner)

implicit val conv = ConfigConvert.fromConfigFieldMapping[ConfWithCamelCase](new WordDelimiterConfigFieldMapping(CamelCaseWordDelimiter, HyphenWordDelimiter))

val conf = ConfigFactory.parseString("""{
  camel-case-int = 1
  camel-case-string = "bar"
  camel-case-conf {
    thisIsAnInt = 3
    thisIsAnotherInt = 10
  }
}""")

conf.to[ConfWithCamelCase]
// returns Success(ConfWithCamelCase(1,bar,ConfWithCamelCaseInner(3,10)))
```